java/org/apache/catalina/valves/SSLValve.java

```diff
@@ -137,7 +149,13 @@ public void invoke(Request request, Response response) throws IOException, Servl
          *       separate lines, the CertificateFactory is tolerant of any
          *       additional whitespace.
          */
-        String headerValue = mygetHeader(request, sslClientCertHeader);
+        String headerValue;
+        String headerEscapedValue = mygetHeader(request, sslClientEscapedCertHeader);
+        if (headerEscapedValue != null) {
+            headerValue = URLDecoder.decode(headerEscapedValue, "ISO-8859-1");
```

Comment:
No, URLDecoder does *not* decode URIs. It has been designed for HTML forms only. Tomcat has a utility to do this properly. Moreover, it should be either UTF-8 or better the encoding from the `server.xml`.

Reply:
Thanks yes I'm aware of what `URLDecoder` does but in this instance it was "good enough" given that we're decoding an ad hoc encoding done by Nginx that is a subset of what `URLDecoder` can decode. Also I had searched in the Tomcat codebase for a utility to do this but didn't find anything.

I searched again deeper and finally found `UDecoder`/`UEncoder` which seem very old and have weird calling conventions but do the job.

Regarding the charset I was using ISO-8859-1 because that's what's used a few lines below when converting the cert to a byte array, but I can use UTF-8 if you prefer. Again this is for an ad hoc Nginx encoding that encodes an ASCII cert anyway so the end result is the same.

The PR has been updated with these changes.

Reply:
Yes, I was referring to the UDecoder. I have nagged last year the usage of URLEncoder and Mark removed it. We should avoid it altogether for concistency. Since URLs are by default UTF-8, you should use UTF-8 unless there is a good reason for that.